### PR TITLE
AWS: Add table and namespace S3 tags

### PR DESCRIPTION
**Pull request by rajarshisarkar**
This change adds table level [S3 Tags](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html) to the objects while writing using `S3FileIO`. Users can pass these opt-in catalog properties to tag the objects in S3:

Spark SQL launch command:
```
sh spark-sql --conf spark.sql.catalog.my_catalog=org.apache.iceberg.spark.SparkCatalog \
    --conf spark.sql.catalog.my_catalog.warehouse=s3://<bucket>/s3-tagging \
    --conf spark.sql.catalog.my_catalog.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog \
    --conf spark.sql.catalog.my_catalog.io-impl=org.apache.iceberg.aws.s3.S3FileIO \
    --conf spark.sql.catalog.my_catalog.s3.write.table-name-tag-enabled=true \
    --conf spark.sql.catalog.my_catalog.s3.write.namespace-name-tag-enabled=true
```

Tags added in S3:

```
aws s3api get-object-tagging --bucket <bucket> --key s3-tagging/data/00000-0-a81d170b-9d71-4b45-886a-390088b2513b-00001.parquet
{
    "TagSet": [
        {
            "Key": "iceberg.table",
            "Value": "test_table_tag"
        },
        {
            "Key": "iceberg.namespace",
            "Value": "test_db"
        }
    ]
}
```

---
cc: @jackye1995 @arminnajafi @singhpk234 @amogh-jahagirdar @xiaoxuandev @yyanyy

### CODE REVIEW

**Comment by danielcweeks** [comment]
I think it would be good to discuss what the use case is for this. I can understand trying to annotate some additional information but I'm not sure that this tracks very well with the evolution of tables. If you rename a table now all of the objects are out of sync.  This also isn't particularly helpful for recovering table data because you actually don't know based on the file what manifests they belong to and that of course can change as well.  

Because iceberg metadata stems from the metadata file all the way down to the physical file, I'm not sure there's any reason to do this that I can think of.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 1]
```diff
@@ -342,11 +372,17 @@
...
+  private boolean s3WriteTableNameTagEnabled;
+  private boolean s3WriteNamespaceTagEnabled;
   private final Map<String, String> s3BucketToAccessPointMapping;
 
   private String glueCatalogId;
   private boolean glueCatalogSkipArchive;
   private boolean glueLakeFormationEnabled;
+  private String tableName;
+  private String namespace;
+
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:385]: nit: extra newline

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 2]
```diff
@@ -388,6 +429,8 @@ public AwsProperties(Map<String, String> properties) {
     this.glueLakeFormationEnabled = PropertyUtil.propertyAsBoolean(properties,
         GLUE_LAKEFORMATION_ENABLED,
         GLUE_LAKEFORMATION_ENABLED_DEFAULT);
+    this.tableName = properties.get(LAKE_FORMATION_TABLE_NAME);
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:432]: we should not get the table and namespace name here, this should be retrieved when creating `newTableOps`, otherwise it's always going to be null

Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: Few doubts:
1. For table level S3 tags, do we need to create new FileIOs for each `newTableOps` call (when LF is not enabled)? The current condition is:
```
protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
    if (catalogProperties != null) {
    ...
    }
```
2. The `tableIdentifier` doesn't have table UUID right now. Should I add the same in the table identifier object?

**Comment by rajarshisarkar** [comment]
> I think it would be good to discuss what the use case is for this. I can understand trying to annotate some additional information but I'm not sure that this tracks very well with the evolution of tables. If you rename a table now all of the objects are out of sync.

Yes, only the table uuid tag would in sync for this scenario. @jackye1995 on the use case.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 3]
```diff
@@ -357,6 +388,18 @@ public class AwsProperties implements Serializable {
    */
   public static final String LAKE_FORMATION_DB_NAME = "lakeformation.db-name";
 
+  /**
+   * Used by {@link GlueCatalog}.
+   * The table name is used as part of table specific catalog properties.
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:393]: nit: as a part of

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 4]
```diff
@@ -357,6 +388,18 @@ public class AwsProperties implements Serializable {
...
+   */
+  public static final String TABLE_NAME = "table-name";
+
+  /**
+   * Used by {@link GlueCatalog}.
+   * The database name is used as part of table specific catalog properties.
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:399]: nit: as a part of

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 5]
```diff
@@ -329,6 +330,36 @@ public class AwsProperties implements Serializable {
...
+  public static final boolean S3_WRITE_TABLE_NAME_TAG_ENABLED_DEFAULT = false;
+
+  /**
+   * Used by {@link S3FileIO} to tag objects when writing. To set, we can pass a catalog property.
+   * <p>
+   * For more details, see https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html
+   * <p>
+   * Example: s3.write.namespace-tag-enabled=true
+   */
+  public static final String S3_WRITE_NAMESPACE_TAG_ENABLED = "s3.write.namespace-tag-enabled";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:360]: should this be `namespace-name`?

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 6]
```diff
@@ -371,11 +414,16 @@ public class AwsProperties implements Serializable {
...
+  private boolean s3WriteTableUuidTagEnabled;
+  private boolean s3WriteTableNameTagEnabled;
+  private boolean s3WriteNamespaceTagEnabled;
   private final Map<String, String> s3BucketToAccessPointMapping;
 
   private String glueCatalogId;
   private boolean glueCatalogSkipArchive;
   private boolean glueLakeFormationEnabled;
+  private String tableName;
+  private String namespace;
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:426]: similarly, namespaceName

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 7]
```diff
@@ -357,6 +388,18 @@ public class AwsProperties implements Serializable {
    */
   public static final String LAKE_FORMATION_DB_NAME = "lakeformation.db-name";
 
+  /**
+   * Used by {@link GlueCatalog}.
+   * The table name is used as part of table specific catalog properties.
+   */
+  public static final String TABLE_NAME = "table-name";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:395]: 1. The `table-name` and `db-name` properties are quite hacky... why not just dynamically add the tags based on catalog tag prefix `s3.write.tags.`? For example, if table name tag enabled, add property `s3.write.tags.table-name` in Glue catalog.
2. the tag names themselves, I think we should probably prefix them by `iceberg.`, for example `iceberg.table-name`, `iceberg.namespace-name`.
3. the variable name should really indicate this is a special tag name, for example `S3_TABLE_SPECIFIC_WRITE_TAGS_TABLE_NAME`.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 8]
```diff
@@ -329,6 +330,36 @@ public class AwsProperties implements Serializable {
...
   public static final boolean S3_DELETE_ENABLED_DEFAULT = true;
 
+  /**
+   * Used by {@link S3FileIO} to tag objects when writing. To set, we can pass a catalog property.
+   * <p>
+   * For more details, see https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html
+   * <p>
+   * Example: s3.write.table-uuid-tag-enabled=true
+   */
+  public static final String S3_WRITE_TABLE_UUID_TAG_ENABLED = "s3.write.table-uuid-tag-enabled";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:340]: Using UUID as tag might be a bit tricker, because when creating table you don't know the table UUID yet when you create the table operation. So I would suggest we add table and namespace name tags first, and add UUID in a separated PR.

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 9]
```diff
@@ -110,13 +110,15 @@ public GlueCatalog() {
   @Override
   public void initialize(String name, Map<String, String> properties) {
     AwsClientFactory awsClientFactory;
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.<String, String>builder().putAll(properties);
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:113]: I think we can make this a `tableSpecificPropertiesBuilder`, and at the end if this builds an empty map it means there is no custom table specific catalog properties, and we should use the else clause.

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 10]
```diff
@@ -197,6 +199,10 @@ protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
               IcebergToGlueConverter.getDatabaseName(tableIdentifier))
           .put(AwsProperties.LAKE_FORMATION_TABLE_NAME,
               IcebergToGlueConverter.getTableName(tableIdentifier))
+          .put(AwsProperties.DB_NAME,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:202]: this is what I meant by we can reuse `s3.write.tags.` property, if we just add tags in that way we directly leverage existing logic.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 6 above)
Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: This is not needed anymore as we are directly leveraging existing logic.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 8 above)
Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: Makes sense to me.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 7 above)
Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: I have made the changes.

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java (hunk 10 above)
Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java]: I have made the changes to leverage the existing logic.

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java (hunk 9 above)
Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java]: [Doubt] If the user passes only the table level tag properties, then we will build the map and according to this logic we would skip the else block. Thereby not initialising `awsClientFactory` and `catalogFileIO` which should not be the intended behaviour.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 11]
```diff
@@ -357,6 +380,18 @@ public class AwsProperties implements Serializable {
    */
   public static final String LAKE_FORMATION_DB_NAME = "lakeformation.db-name";
 
+  /**
+   * Used by {@link GlueCatalog}.
+   * The table name is used as a part of table specific catalog properties.
+   */
+  public static final String S3_TABLE_SPECIFIC_WRITE_TAGS_TABLE_NAME = "iceberg.table-name";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:387]: looks like this should be moved to under `S3_WRITE_TAGS_PREFIX`? And we can have a simpler name like `S3_TAG_TABLE_NAME`

Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: Thanks, I have made the change.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 12]
```diff
@@ -370,6 +371,44 @@ public class AwsProperties implements Serializable {
...
+   * https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html
+   *
+   * <p>Example: s3.write.namespace-name-tag-enabled=true
+   */
+  public static final String S3_WRITE_NAMESPACE_TAG_ENABLED = "s3.write.namespace-name-tag-enabled";
+
+  public static final boolean S3_WRITE_NAMESPACE_TAG_ENABLED_DEFAULT = false;
+
+  /**
+   * Used by {@link GlueCatalog}. The table name is used as a part of table specific catalog
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:401]: we should note this is the tag name that will be used by `{@link S3_WRITE_TAGS_PREFIX}`

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java [hunk 13]
```diff
@@ -355,4 +355,8 @@ void cleanupMetadataAndUnlock(CommitStatus commitStatus, String metadataLocation
       }
     }
   }
+
+  public Map<String, String> tableCatalogProperties() {
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:359]: looks like this is only for testing purpose? can we use spy for that instead of exposing it?

File: aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java [hunk 14]
```diff
@@ -585,4 +585,32 @@ public void testValidateIdentifierSkipNameValidation() {
         CATALOG_NAME, WAREHOUSE_PATH, props, glue, LockManagers.defaultLockManager(), null);
     Assert.assertEquals(glueCatalog.isValidIdentifier(TableIdentifier.parse("db-1.a-1")), true);
   }
+
+  @Test
+  public void testTableLevelS3TagProperties() {
+    Map<String, String> properties =
+        ImmutableMap.of(
+            "s3.write.table-name-tag-enabled",
```
Comment (jackye1995) [inline aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java:593]: nit: can import from `AwsProperties`

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 15]
```diff
@@ -199,20 +200,31 @@ void initialize(
...
-                  AwsProperties.LAKE_FORMATION_DB_NAME,
-                  IcebergToGlueConverter.getDatabaseName(
-                      tableIdentifier, awsProperties.glueCatalogSkipNameValidation()))
-              .put(
-                  AwsProperties.LAKE_FORMATION_TABLE_NAME,
-                  IcebergToGlueConverter.getTableName(
-                      tableIdentifier, awsProperties.glueCatalogSkipNameValidation()));
+          ImmutableMap.<String, String>builder().putAll(catalogProperties);
+
+      if (isTableLevelTagEnabled(catalogProperties)) {
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:205]: we already have `awsProperties` here, no need to parse property map again

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 16]
```diff
@@ -113,10 +113,11 @@ public void initialize(String name, Map<String, String> properties) {
     this.catalogProperties = ImmutableMap.copyOf(properties);
     AwsClientFactory awsClientFactory;
     FileIO catalogFileIO;
-    if (PropertyUtil.propertyAsBoolean(
-        properties,
-        AwsProperties.GLUE_LAKEFORMATION_ENABLED,
-        AwsProperties.GLUE_LAKEFORMATION_ENABLED_DEFAULT)) {
+    if (isTableLevelTagEnabled(properties)) {
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:116]: why is this needed?

**Comment by jackye1995** [comment]
Thanks for updating this, could you also add an integration test to verify files written to a table are properly tagged in s3?

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java (hunk 16 above)
Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java]: Shouldn't be needed, I missed that `catalogProperties` was initialised as `ImmutableMap.copyOf(properties);`

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 17]
```diff
@@ -394,6 +395,48 @@ public class AwsProperties implements Serializable {
...
+  /**
+   * Used by {@link GlueCatalog} to tag objects when writing. To set, we can pass a catalog
+   * property.
+   *
+   * <p>For more details, see
+   * https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html
+   *
+   * <p>Example: s3.write.table-name-tag-enabled=true
+   */
+  public static final String S3_WRITE_TABLE_NAME_TAG_ENABLED = "s3.write.table-name-tag-enabled";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:407]: I am wondering if we should simplify these config names to `s3.write.table-tag-enabled`, `s3.write.namespace-tag-enabled`, `iceberg.table`, `iceberg.namespace`, what do you think?

Comment (rajarshisarkar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: Yes, seems clean to me. I have made the changes.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 18]
```diff
@@ -394,6 +395,48 @@ public class AwsProperties implements Serializable {
...
+
+  public static final boolean S3_WRITE_NAMESPACE_TAG_ENABLED_DEFAULT = false;
+
+  /**
+   * Tag name that will be used by {@link #S3_WRITE_TAGS_PREFIX} when {@link
+   * #S3_WRITE_TABLE_TAG_ENABLED} is enabled
+   *
+   * <p>Example: iceberg.table=tableName
+   */
+  public static final String S3_TAG_TABLE = "iceberg.table";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:430]: what about `S3_TAG_ICEBERG_TABLE`? just to match the value

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 19]
```diff
@@ -498,6 +541,8 @@ public class AwsProperties implements Serializable {
   private ObjectCannedACL s3FileIoAcl;
   private boolean isS3ChecksumEnabled;
   private final Set<Tag> s3WriteTags;
+  private boolean s3WriteTableNameTagEnabled;
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:544]: can we match the config name and name this `s3WriteTableTagEnabled`?

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 20]
```diff
@@ -856,6 +909,22 @@ public Set<Tag> s3WriteTags() {
     return s3WriteTags;
   }
 
+  public boolean s3WriteTableNameTagEnabled() {
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:912]: same here for the method name matching config name

**Review by jackye1995 — APPROVED** [review]
looks good to me!

**Comment by jackye1995** [comment]
cc @singhpk234 @amogh-jahagirdar for any comment

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 21]
```diff
@@ -199,20 +199,33 @@ void initialize(
...
-                  IcebergToGlueConverter.getTableName(
-                      tableIdentifier, awsProperties.glueCatalogSkipNameValidation()));
+          ImmutableMap.<String, String>builder().putAll(catalogProperties);
 
-      if (awsProperties.glueLakeFormationEnabled()) {
+      if (awsProperties.s3WriteTableTagEnabled()) {
+        tableSpecificCatalogPropertiesBuilder.put(
+            AwsProperties.S3_WRITE_TAGS_PREFIX.concat(AwsProperties.S3_TAG_ICEBERG_TABLE),
+            IcebergToGlueConverter.getTableName(
+                tableIdentifier, awsProperties.glueCatalogSkipNameValidation()));
```
Comment (singhpk234) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:208]: [minor] can extract this into local vars to be reused in both places.

**Review by singhpk234 — APPROVED** [review]
LGTM as well !!

**Review by amogh-jahagirdar — APPROVED** [review]
Overall LGTM, thanks!

**Comment by jackye1995** [comment]
Thanks for finishing the work @rajarshisarkar ! and thanks @amogh-jahagirdar and @singhpk234 for the reviews!
